### PR TITLE
Migrate to Spark 2.0

### PR DESCRIPTION
**Pull request by liancheng**
This PR migrates spark-redshift to Spark 2.0.

It is marked as WIP because it depends on https://github.com/databricks/spark-avro/pull/137.

### CODE REVIEW

File: project/SparkRedshiftBuild.scala [hunk 1]
```diff
@@ -44,9 +44,9 @@ object SparkRedshiftBuild extends Build {
       organization := "com.databricks",
       scalaVersion := "2.10.5",
       crossScalaVersions := Seq("2.10.5", "2.11.7"),
-      sparkVersion := "1.4.1",
+      sparkVersion := "2.0.0-SNAPSHOT",
```
Comment (clockfly) [inline project/SparkRedshiftBuild.scala]: Probably we should change version in title to "Spark 2.0.0-SNAPSHOT".

**Comment by clockfly** [comment]
Travis file is not updated, now the build fails.
https://github.com/databricks/spark-redshift/blob/master/.travis.yml

**Comment by clockfly** [comment]
Should we use a new SNAPSHOT version considering 0.6.1-SNAPSHOT is marked 6 months ago and this PR introduces an incompatible change? Probably 0.6.2-SNAPSHOT?

**Comment by clockfly** [comment]
Should we also update README.md? Since this introduces an incompatible change, and in the README, there is section like:
```
This library requires Apache Spark 1.4+ and Amazon Redshift 1.0.963+.
```

**Comment by liancheng** [comment]
@clockfly Thanks for the review, I haven't updated the build stuff since it's still blocked by https://github.com/databricks/spark-avro/pull/137, and Travis build won't pass anyway. Will update them later.

**Comment by liancheng** [comment]
Updated version information and dependencies. Now we are using newly released spark-avro 3.0.0-preview2. This is ready for review now.

**Comment by liancheng** [comment]
@JoshRosen Would you mind to help review this one? Thanks!

**Comment by JoshRosen** [comment]
In order to trigger the end-to-end integration tests, you need to push this code into a branch in the `databricks/spark-redshift` repo (since Travis won't decrypt credentials for pull-request jobs). I'm going to push a copy of this to a temporary branch to see if the tests pass there.

This looks good to me, but I'm wondering if we're planning to make a new release for Spark 1.x in the near term. If we do, then it's going to mean that we'll have to make a separate 1.x branch and merge outstanding PRs twice, once into the master branch and once into the 1.x maintenance branch. That's not a huge deal, though.

File: README.md [hunk 2]
```diff
@@ -22,22 +22,22 @@ This library is more suited to ETL than interactive queries, since large amounts
 
 ## Installation
 
-This library requires Apache Spark 1.4+ and Amazon Redshift 1.0.963+.
+This library requires Apache Spark 2.0+ and Amazon Redshift 1.0.963+.
```
Comment (JoshRosen) [inline README.md]: Given that this is going to be the README when people browse to the repository on GitHub, we should probably have a pointer to a branch containing the Spark 1.x version of the code or a table listing the releases for the Spark 1.x line.

**Comment by JoshRosen** [comment]
I'm going to try to help review and merge the other outstanding PRs right now so that we can hopefully make a new 1.x release.

**Comment by JoshRosen** [comment]
Status update: merged a number of open PRs and am on track to wrap up #157 and then will cut a release. In case this merge-conflicts I'll take care of rebasing.

File: README.md (hunk 2 above)
Comment (liancheng) [inline README.md]: Should I do this after making the branch for Spark 1.x?

**Comment by liancheng** [comment]
@JoshRosen Thanks for the review!

File: README.md (hunk 2 above)
Comment (JoshRosen) [inline README.md]: Created a new branch at https://github.com/databricks/spark-redshift/tree/branch-1.x for holding the 1.x code.

File: README.md [hunk 3]
```diff
@@ -22,22 +22,22 @@ This library is more suited to ETL than interactive queries, since large amounts
...
 
 You may use this library in your applications with the following dependency information:
 
 **Scala 2.10**
 ```
 groupId: com.databricks
 artifactId: spark-redshift_2.10
-version: 0.6.0
+version: 0.7.0
```
Comment (JoshRosen) [inline README.md]: I'd change these versions to `2.0.0-SNAPSHOT`. Also, make sure to update `version.sbt` in the root of this repo.

File: src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala [hunk 4]
```diff
@@ -84,7 +84,7 @@ private[redshift] case class RedshiftRelation(
   // In Spark 1.6+, this method allows a data source to declare which filters it handles, allowing
```
Comment (JoshRosen) [inline src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala]: This comment can now be safely removed.

**Comment by yhuai** [comment]
@liancheng Let's also update README. Thanks!

**Comment by JoshRosen** [comment]
Since Spark 2.0+ only supports Hadoop 2.x, we can remove the reflection hack at https://github.com/databricks/spark-redshift/blob/cba5eee1ab79ae8f0fa9e668373a54d2b5babf6b/src/main/scala/com/databricks/spark/redshift/RedshiftInputFormat.scala#L106

Similarly, we can remove the reflection at https://github.com/databricks/spark-redshift/blob/5bc5fabc68f3972c576cc154ddaa1f15ad61f09d/src/main/scala/com/databricks/spark/redshift/RedshiftJDBCWrapper.scala#L97

**Comment by liancheng** [comment]
@JoshRosen @yhuai Thanks for the review! Comments addressed. Shall we also remove deprecated methods like those defined in `RedshiftContext`? I can do it in another PR.

**Comment by JoshRosen** [comment]
I pushed this PR's changes to a branch in the `databricks/spark-redshift` repository so that the integration tests could run and it looks like they're all failing due to changes in TestHive: https://travis-ci.org/databricks/spark-redshift/jobs/144786105

e.g.:

```
info] Exception encountered when attempting to run a suite with class name: com.databricks.spark.redshift.DecimalIntegrationSuite *** ABORTED ***
[info]   java.lang.NullPointerException:
[info]   at org.apache.spark.sql.hive.test.TestHiveSparkSession.getHiveFile(TestHive.scala:183)
[info]   at org.apache.spark.sql.hive.test.TestHiveSparkSession.<init>(TestHive.scala:214)
[info]   at org.apache.spark.sql.hive.test.TestHiveSparkSession.<init>(TestHive.scala:122)
[info]   at org.apache.spark.sql.hive.test.TestHiveContext.<init>(TestHive.scala:77)
[info]   at com.databricks.spark.redshift.IntegrationSuiteBase$class.beforeEach(IntegrationSuiteBase.scala:122)
[info]   at com.databricks.spark.redshift.DecimalIntegrationSuite.beforeEach(DecimalIntegrationSuite.scala:25)
[info]   at org.scalatest.BeforeAndAfterEach$class.beforeEach(BeforeAndAfterEach.scala:154)
[info]   at com.databricks.spark.redshift.DecimalIntegrationSuite.beforeEach(DecimalIntegrationSuite.scala:25)
[info]   at org.scalatest.BeforeAndAfterEach$class.beforeEach(BeforeAndAfterEach.scala:173)
[info]   at com.databricks.spark.redshift.DecimalIntegrationSuite.beforeEach(DecimalIntegrationSuite.scala:25)
[info]   at org.scalatest.BeforeAndAfterEach$class.runTest(BeforeAndAfterEach.scala:253)
[info]   at com.databricks.spark.redshift.DecimalIntegrationSuite.runTest(DecimalIntegrationSuite.scala:25)
[info]   at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
```

**Comment by JoshRosen** [comment]
Let's remove that deprecated stuff in a separate PR, by the way.

**Comment by liancheng** [comment]
The NPE was thrown when loading test tables in `TestHiveContext` constructor because of missing testing data files. PR apache/spark#14005 added a flag to disable loading test tables. However, this change happened after 2.0.0-RC1 and was only included in 2.0.1-SNAPSHOT. Thus I updated Spark version to 2.0.1-SNAPSHOT so that we can use this flag to avoid the NPE.

**Comment by JoshRosen** [comment]
@liancheng, can we use 2.0 RC4?

Also, it looks like a filter pushdown test is failing because the assert assumed that the physical plan would push all filters to Redshift while it appears that "NOT NULL" isn't being pushed:

```
RedshiftRelation implements Spark 1.6+'s unhandledFilters API *** FAILED ***
[info]   Filter should have been eliminated; plan is:
[info]   Filter isnotnull(testbool#814)
[info]   +- Scan RedshiftRelation("PUBLIC"."test_table_438264183817572114") [testbool#814] PushedFilters: [IsNotNull(testbool), EqualTo(testbool,true)] (RedshiftIntegrationSuite.scala:289)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:495)
[info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1555)
[info]   at org.scalatest.Assertions$class.fail(Assertions.scala:1328)
[info]   at org.scalatest.FunSuite.fail(FunSuite.scala:1555)
[info]   at com.databricks.spark.redshift.RedshiftIntegrationSuite$$anonfun$10$$anonfun$apply$mcV$sp$4.apply(RedshiftIntegrationSuite.scala:289)
[info]   at com.databricks.spark.redshift.RedshiftIntegrationSuite$$anonfun$10$$anonfun$apply$mcV$sp$4.apply(RedshiftIntegrationSuite.scala:288)
[info]   at scala.Option.foreach(Option.scala:236)
```

https://github.com/databricks/spark-redshift/blob/a1749cb8dd3b064a930c19f8cd12e557a274b1aa/src/it/scala/com/databricks/spark/redshift/RedshiftIntegrationSuite.scala#L288

To fix this, I think we can add a pushdown for IsNutNull in https://github.com/databricks/spark-redshift/blob/a1749cb8dd3b064a930c19f8cd12e557a274b1aa/src/main/scala/com/databricks/spark/redshift/FilterPushdown.scala#L74

**Comment by liancheng** [comment]
@JoshRosen I tried to use RCs before. But RCs are not releases and no artifacts would be published to the central Maven repo. That's why I had to stick to a snapshot release. A preview release should also work but the latest preview release is too old. I'm fixing the filter push-down issue.

**Comment by JoshRosen** [comment]
LGTM. As long as build against something that's binary-compatible with the final 2.0.0 release I think we'll be fine, especially since we'll publish a non-preview release of `spark-redshift` as soon as the final 2.0.0 vote passes.

**Comment by liancheng** [comment]
@JoshRosen Thanks!

**Pull request closed without merging**